src/vault.c

```diff
@@ -0,0 +1,329 @@
+        return NULL;
+    }
+    //TODO: ? connection requests will be queued
+    if (listen(server_fd, 10) < 0) {
+        pgagroal_log_error("Listen failed");
+        return NULL;
+    }
+
+    printf("HTTP server is listening on port %d...\n", HTTP_PORT);
+
+    while (1) {
+        client_fd = accept(server_fd, (struct sockaddr *)&client_addr, &client_len);
+        if (client_fd < 0) {
+            pgagroal_log_error("Accept failed");
+            continue;
+        }
+
+        handle(client_fd, NULL, 1);  // 1 indicates HTTP
+    }
+
+    close(server_fd);
+    return NULL;
+}
+
+static void *https_thread_func(void *arg) {
+    int server_fd, client_fd;
+    struct sockaddr_in server_addr, client_addr;
+    socklen_t client_len = sizeof(client_addr);
+
+    server_fd = socket(AF_INET, SOCK_STREAM, 0);
+    if (server_fd < 0) {
+        pgagroal_log_error("Cannot create socket");
+        return NULL;
+    }
+
+    memset(&server_addr, 0, sizeof(server_addr));
+    server_addr.sin_family = AF_INET;
+    server_addr.sin_addr.s_addr = INADDR_ANY;
+    server_addr.sin_port = htons(HTTPS_PORT);
+
+    if (bind(server_fd, (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0) {
+        pgagroal_log_error("Bind failed");
+        return NULL;
+    }
+
+    if (listen(server_fd, 10) < 0) {
+        pgagroal_log_error("Listen failed");
+        return NULL;
+    }
+
+    printf("HTTPS server is listening on port %d...\n", HTTPS_PORT);
+
+    while (1) {
+        client_fd = accept(server_fd, (struct sockaddr *)&client_addr, &client_len);
+        if (client_fd < 0) {
+            pgagroal_log_error("Accept failed");
+            continue;
+        }
+
+        SSL *ssl = SSL_new(ssl_ctx);
+        SSL_set_fd(ssl, client_fd);
+
+        if (SSL_accept(ssl) <= 0) {
+            pgagroal_log_error("pgagroal_vault: ssl filed to accept");
+        } else {
+           handle(client_fd, ssl, 0);  // 0 indicates HTTPS
+        }
+
+        // SSL_free(ssl);
+    }
+
+    close(server_fd);
+    return NULL;
+}
+
+int main(int arvc, char** argv) {
+    initialize_ssl();
+    ssl_ctx = create_ssl_context();
+    configure_ssl_context(ssl_ctx);
+
+    printf("creating thread\n");
+
+    client_socket = socket(AF_INET, SOCK_STREAM, 0);
+    if (client_socket == -1) {
+        perror("Socket creation error");
+        exit(EXIT_FAILURE);
+    }
+    
+    struct sockaddr_in serverAddress;
+    serverAddress.sin_family = AF_INET;
+    serverAddress.sin_port = htons(6789);
+    serverAddress.sin_addr.s_addr = inet_addr("127.0.0.1");
+    
+    if (connect(client_socket, (struct sockaddr *)&serverAddress, sizeof(serverAddress)) == -1) {
+        perror("Connection error");
+        close(client_socket);
+        exit(EXIT_FAILURE);
+    }
+
+    pthread_t http_thread, https_thread;
+
+    int result = pthread_create(&http_thread, NULL, http_thread_func, NULL);
+    if(result != 0) {
+        printf("something wrong\n");
```

Comment:
warnx

Reply:
Thank you for your feedback. I am aware that there are many flaws in the code I submitted in the PR. I opened this PR to discuss my implementation with you all. Once we have finalized how to implement these features, I will close this PR and submit a clean one.